enhancements/authentication/login-logout-events.md

```diff
@@ -0,0 +1,308 @@
+
+```json
+{
+  "kind": "Event",
+  "apiVersion": "audit.k8s.io/v1",
+  "level": "RequestResponse",
+...
+  "verb": "delete",
+...
+  "responseObject": {
+    "kind": "Status",
+    "apiVersion": "v1",
+    "metadata": {},
+    "status": "Success",
+    "details": {
+      "name": "sha256~sometoken",
+      "group": "oauth.openshift.io",
+      "kind": "oauthaccesstokens",
+      "uid": "444845f0-1348-4ab1-95be-d0644a6bd11c"
+    }
+  },
+...
+}
+```
+
+To log the complete object we have to set `ReturnDeletedObject: true` on the oauth-apiserver REST Store: https://github.com/openshift/oauth-apiserver/blob/6e0f92194d5a25728c826fefb2d99c5e88ebb5e5/pkg/oauth/apiserver/registry/oauthaccesstoken/etcd/etcd.go#L30-L46.
+
+This will allow to introspect the deleted user:
+
+```json
+{
+   "kind": "Event",
+   "apiVersion": "audit.k8s.io/v1",
+   "level": "RequestResponse",
+...
+    "verb": "delete",
+...
+    "responseObject": {
+      "kind": "OAuthAccessToken",
+      "apiVersion": "oauth.openshift.io/v1",
+      "metadata": {
+...
+        "userName": "exampleuser",
+      },
+      "userUID": "123",
+      "authorizeToken": "sha256~sometoken"
+    },
+...
+  }
+```
+
+### Event construction
+
+In all cases above an [Info](https://github.com/openshift/oauth-server/blob/961295a3151da3fff9daa5ea0ab8c0bf92a7d7e6/vendor/k8s.io/apiserver/pkg/authentication/user/user.go#L20) entity is available with user information which can be used to fill audit event data.
+
+### Integration in must-gather
+
+The existing `oc adm must-gather -- /usr/bin/gather_audit_logs` tool will also fetch oauth-server audit logs.
+
+### DoS mitigation
+
+To prevent flooding audit event logs we rely on the existing default apiserver handler chain which is used to serve oauth-server http resources.
+
+Currently, the default configuration of apiserver's `MaxRequestsInFlight` (i.e. GET requests) and `MaxMutatingRequestsInFlight` (i.e. POST requests) are used as part of the [default configuration](https://github.com/openshift/oauth-server/blob/961295a3151da3fff9daa5ea0ab8c0bf92a7d7e6/vendor/k8s.io/apiserver/pkg/server/config.go#L330-L331) to protect against audit event flooding.
+
+### Graduation Criteria
+
+This feature is planned for OpenShift 4.10.
+
+#### Dev Preview -> Tech Preview
+
+- Audit logging should work.
+
+#### Tech Preview -> GA
+
+- No sensitive data is logged.
+- Verbosity should be acceptable.
+
+#### Removing a deprecated feature
+
+N/A
+
+### Upgrade / Downgrade Strategy
+
+As we are introducing a new property with default values, there are not expected risks for upgrading.
+
+Downgrading would not cause any data loss or inconsistencies as the property is a simple feature flag to toggle audit logging.
+
+### Version Skew Strategy
+
+All components will be developed with the concern of version skew in mind. So they will handle the missing of the `Audit` property.
+
+### Test Plan
+
+Ideally we test our feature with:
+
+- unit tests,
+- e2e tests in origin and
```

Comment:
Note to future devs: https://github.com/openshift/origin/blob/master/test/extended/cluster/audit.go